Azure/sonic-buildimage
# [Mellanox] Modified Platform API to support all firmware updates in single boot

**DEPENDS ON https://github.com/Azure/sonic-platform-common/pull/254 AND https://github.com/Azure/sonic-buildimage/pull/9607 DO NOT MERGE WITHOUT**

#### Why I did it

Requirements from Microsoft for `fwutil update all` state that all firmwares which support this upgrade flow must support upgrade within a single boot cycle. This conflicted with a number of Mellanox upgrade flows which have been revised to safely meet this requirement. 

#### How I did it

- Added `--no-power-cycle` flags to SSD and ONIE firmware scripts
- Modified Platform API to call firmware upgrade flows with this new flag during `fwutil update all` 
- Added a script to our reboot plugin to handle installing firmwares in the correct order with prior to reboot

#### How to verify it

1. Populate `platform_components.json` with firmware for CPLD / BIOS / ONIE / SSD
2. Execute `fwutil update all fw --boot cold`
_CPLD will burn /  ONIE and BIOS images will stage / SSD will schedule for reboot_
3. Reboot the switch
_SSD will install / CPLD will refresh / switch will power cycle into ONIE_
_ONIE installer will upgrade ONIE and BIOS / switch will reboot back into SONiC_
4. In SONiC run `fwutil show status` to check that all firmware upgrades were successful 

#### Which release branch to backport (provide reason below if selected)

None

#### Description for the changelog
[mellanox] Modified Platform API to support all firmware updates in single boot

#### A picture of a cute animal (not mandatory but encouraged)
![red-panda-full-body](https://user-images.githubusercontent.com/5898707/146974803-e9a89e0f-c36b-49bc-ae92-1ec122802c15.jpg)
